epitator/structured_data_annotator.py

```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python
+from __future__ import absolute_import
+from .annotator import Annotator, AnnoTier, AnnoSpan
+import re
+import pyparsing as pypar
+
+
+def word_token_regex(disallowed_delimiter):
+    return pypar.Regex(r"[^\s\n" + re.escape(disallowed_delimiter) + r"]+")
+
+
+pypar.ParserElement.setDefaultWhitespaceChars(" \t")
+table_parser = pypar.NoMatch()
+table_cell_separators = ["|", "/", ","]
+for separator in table_cell_separators:
+    value = pypar.Combine(word_token_regex(separator) * (1, 10), joinString=' ', adjacent=False)
+    row = pypar.Group(pypar.Optional(separator).suppress() +
+                      (value + pypar.Literal(separator).suppress()) * (2, None) +
+                      value + pypar.Optional(separator).suppress() + pypar.Literal(
+        "\n").suppress() + pypar.Optional("\n").suppress())
+    # Search for two or more rows of similar data
+    table_parser ^= row * (2, None)
+
+key_value_separators = [":", "-", ">"]
+key_value_list_parser = pypar.NoMatch()
+for separator in key_value_separators:
+    value = pypar.Combine(word_token_regex(separator) * (1, 10), joinString=' ', adjacent=False)
+    row = pypar.Group(value + pypar.Literal(separator).suppress() + value + pypar.Literal(
+        "\n").suppress() + pypar.Optional("\n").suppress())
+    key_value_list_parser ^= row * (2, None)
+
+
+class StructuredDataAnnotator(Annotator):
+    """
+    Annotates tables and key value lists embedded in documents.
+    """
+
+    def annotate(self, doc):
+        spans = []
+        for token, start, end in table_parser.scanString(doc.text):
+            spans.append(AnnoSpan(start, end, doc, "table", metadata={
+                "type": "table",
+                "data": [[value for value in row] for row in token]
+            }))
+        for token, start, end in key_value_list_parser.scanString(doc.text):
+            spans.append(AnnoSpan(start, end, doc, "keyValuePairs", metadata={
+                "type": "keyValuePairs",
+                "data": {
+                    pair[0]: pair[1]
+                    for pair in token
+                }
+            }))
+        return {'structuredData': AnnoTier(spans)}
```

Comment:
This seems like a good first pass. One thing to note: There's another example of a key-value pair in the [Slack discussion of exported annotations]()https://eha.slack.com/archives/G0P7PPUAW/p1511302007000117):
>The number of ill people identified in each state with the outbreak strain is as follows: Alabama (2), Arkansas (1), Connecticut (4), District of Columbia (2), Georgia (4), Illinois (8), Louisiana (2), Maryland (8), Massachusetts (4), Mississippi (1), Missouri (1), New Jersey (6), New York (23), North Carolina (2), Pennsylvania (2), Rhode Island (4), South Carolina (3), Texas (3), Virginia (5), and Wisconsin (8).

I wonder if it'd be useful broadly to also flag non-natural-language regions so that our NLP algorithms don't go haywire over them. I'm thinking of things in addition to tables and key-value lists — things like front matter etc.

But for now — I think this is a good first pass, and I'm interested to see how it operates. We should probably not necessarily use it to create Incidents yet, b/c I'm sure that the way we usually assign dates and locations and other metadata to them won't necessarily apply to this stuff.